paramiko/pkey.py

```diff
@@ -52,36 +178,121 @@ def unpad(data):
+    kdfname = message.get_text()
+    kdfoptions = message.get_binary()
+    num_keys = message.get_int()
+
+    if num_keys > 1:
+        raise SSHException("unsupported: private keyfile has multiple keys")
+
+    public_data = message.get_binary()
+    privkey_blob = message.get_binary()
+
+    pub_keytype = Message(public_data).get_text()
+
+    if kdfname == "none":
+        if kdfoptions or cipher != "none":
+            raise SSHException("Invalid key options for kdf 'none'")
+        private_data = privkey_blob
+
+    elif kdfname == "bcrypt":
+        if not password:
+            raise PasswordRequiredException("Private key file is encrypted")
+        if cipher == 'aes256-cbc':
+            mode = modes.CBC
+        elif cipher == 'aes256-ctr':
+            mode = modes.CTR
+        else:
+            raise SSHException("unknown cipher '%s' used in private key file" % cipher)
+
+        kdf = Message(kdfoptions)
+        salt = kdf.get_binary()
+        rounds = kdf.get_int()
+
+        # run bcrypt kdf to derive key and iv/nonce (desired_key_bytes = 32 + 16 bytes)
+        key_iv = bcrypt.kdf(b(password), salt, 48, rounds, ignore_few_rounds=True)
+        key = key_iv[:32]
+        iv = key_iv[32:]
+        # decrypt private key blob
+        decryptor = Cipher(algorithms.AES(key), mode(iv), default_backend()).decryptor()
+        private_data = decryptor.update(privkey_blob) + decryptor.finalize()
+
+    else:
+        raise SSHException("unknown cipher or kdf used in private key file")
+
+    # Unpack private key and verify checkints
+    priv_msg = Message(private_data)
+    checkint1 = priv_msg.get_int()
+    checkint2 = priv_msg.get_int()
+    if checkint1 != checkint2:
+        raise SSHException('OpenSSH private key file checkints do not match')
+
+    # FIXME: Is this correct?
+    keytype = priv_msg.get_text()
+    if pub_keytype != keytype:
+        raise SSHException("Inconsistent key types for public and private parts")
+
+    keydata = priv_msg.get_remainder()
+    return keytype, unpad(keydata)
+
+
 class PKey(object):
     """
     Base class for public keys.
     """
 
-    # known encryption types for private key files:
-    _CIPHER_TABLE = {
-        'AES-128-CBC': {
-            'cipher': algorithms.AES,
-            'keysize': 16,
-            'blocksize': 16,
-            'mode': modes.CBC
-        },
-        'AES-256-CBC': {
-            'cipher': algorithms.AES,
-            'keysize': 32,
-            'blocksize': 16,
-            'mode': modes.CBC
-        },
-        'DES-EDE3-CBC': {
-            'cipher': algorithms.TripleDES,
-            'keysize': 24,
-            'blocksize': 8,
-            'mode': modes.CBC
-        },
-    }
     FORMAT_ORIGINAL = 1  # PKCS#1 for RSA or SEC1 for EC
     FORMAT_OPENSSH = 2  # newer "openssh-key-v1" format
-    BEGIN_TAG = re.compile(r'^-{5}BEGIN (RSA|DSA|EC|OPENSSH) PRIVATE KEY-{5}\s*$')
-    END_TAG = re.compile(r'^-{5}END (RSA|DSA|EC|OPENSSH) PRIVATE KEY-{5}\s*$')
```

Comment:
`BEGIN_TAG` and `END_TAG` are the only non-underscore variables that I changed in a backwards-incompatible manner.

I really doubt anyone is relying on these in ther own code, what do you think?

Reply:
yeah, it is extremely unlikely that anyone would be referencing these
👍